src/lazy/encoder/binary/v1_1/container_writers.rs

```diff
@@ -120,60 +122,107 @@ pub struct BinaryStructWriter_1_1<'value, 'top> {
+    pub(crate) fn new_delimited(container_writer: BinaryContainerWriter_1_1<'value, 'top>) -> Self {
+        Self {
+            // Delimited structs always use FlexSym encoding.
+            flex_uint_encoding: false,
+            container_writer,
         }
     }
 
     pub(crate) fn buffer(&mut self) -> &'_ mut BumpVec<'top, u8> {
         self.container_writer.buffer
     }
 
+    #[inline]
     pub fn write<A: AsRawSymbolTokenRef, V: WriteAsIon>(
         &mut self,
         name: A,
         value: V,
     ) -> IonResult<&mut Self> {
         use RawSymbolTokenRef::*;
 
+        match (self.flex_uint_encoding, name.as_raw_symbol_token_ref()) {
+            // We're already in FlexSym encoding mode
+            (false, _) => self.write_flex_sym_field(name, value),
+            // We're still in FlexUInt encoding mode, but this value requires FlexSym encoding
+            (_, Text(_)) | (_, SymbolId(0)) =>
+            // This can only happen up to once inside a length-prefixed struct
+            {
+                self.enable_flex_sym_and_write_field(name, value)
+            }
+            // We're in FlexUInt encoding mode and can write this field without switching modes
+            (_, SymbolId(sid)) => self.write_flex_uint_field(sid, value),
+        }
+    }
+
+    #[inline]
+    fn write_flex_uint_field<V: WriteAsIon>(
+        &mut self,
+        name: SymbolId,
+        value: V,
+    ) -> IonResult<&mut Self> {
+        FlexUInt::encode_u64(self.buffer(), name as u64);
+        self.container_writer.write(value)?;
+        Ok(self)
+    }
+
+    #[inline(never)]
+    fn enable_flex_sym_and_write_field<A: AsRawSymbolTokenRef, V: WriteAsIon>(
+        &mut self,
+        name: A,
+        value: V,
+    ) -> IonResult<&mut Self> {
+        if self.flex_uint_encoding {
+            // This is the first time we're writing a FlexSym field. Emit a FlexUInt 0 to tell
+            // readers that we're switching from FlexUInt to FlexSym.
+            self.buffer().push(0x01);
+            self.flex_uint_encoding = false;
+        }
+        self.write_flex_sym_field(name, value)
+    }
+
+    pub fn write_flex_sym_field<A: AsRawSymbolTokenRef, V: WriteAsIon>(
+        &mut self,
+        name: A,
+        value: V,
+    ) -> IonResult<&mut Self> {
+        use RawSymbolTokenRef::*;
+
         // Write the field name
         match name.as_raw_symbol_token_ref() {
             SymbolId(0) => {
-                self.enable_flex_sym_encoding();
                 // Encoding `$0` requires a zero byte to indicate an opcode follows,
-                // and then opcode 0x70, which indicates symbol ID 0.
-                self.buffer().extend_from_slice(&[0, 0x70]);
-            }
-            SymbolId(sid) if self.flex_uint_encoding => {
-                FlexUInt::write_u64(self.buffer(), sid as u64)?;
+                // and then opcode 0x90, which indicates symbol ID 0.
+                cold_path! {
```

Comment:
> Question—while this one instance of using ice/cold_path might not make much difference on its own, do we need to be concerned about this macro inflating the compiled code?

I don't think we do. You can take a look at [the code the macro expands to](https://github.com/zslayton/ice_code/blob/feb4d3cde483c53f7c1e693a4be401fff14c821b/src/lib.rs#L18-L23), which is pretty trivial.

> I.e.: Is cold_path! going to inflate the size of the compiled code as compared to something like having a function like this so that ice/cold_path doesn't generate a new function for both the 0x80 and 0x90 cases?

I don't have data to point to, but I would expect the macro to actually reduce the size of the final binary by a small amount. It does introduce new functions, but it also inherently reduces the number of places that the content of those functions would appear if they were being inlined.

For example, consider the `write_symbol` method on the v1.0 binary writer:

https://github.com/amazon-ion/ion-rust/blob/859105f007538a74b4ef28b36eeac59ff370ef6a/src/lazy/encoder/binary/v1_0/value_writer.rs#L231-L238

If we wrap the error handling in the `Text` case with `cold_path!`, we introduce a new function in the resulting binary. However, we also prevent `String` formatting machinery from appearing in every location in which `write_symbol` is called. On balance, I'd expect this to do more help than harm.

Adding a `cold_extend_from_slice` method gets a lot of the benefit, but requires the caller to push one of the slice literals onto the stack before calling it, resulting in a minor size increase.

I'd be somewhat surprised if either approach made a material difference in terms of bloat though.